### PR TITLE
DER-124 - Need to address gaps with respect to the CFI representation of swaps

### PR DESCRIPTION
**Pull request by ElisaKendall**
Signed-off-by: Elisa Kendall <ekendall@thematix.com>

## Description

Added the concept of a rates swap and the corresponding rate-based leg to the swaps ontology and integrated it with the relevant subordinate ontologies to facilitate mapping to the CFI standard

Fixes: #1727 / DER-124

## Checklist:

- [x] I'm familiar with the [FIBO developer quide](https://github.com/edmcouncil/fibo/blob/master/CONTRIBUTING.md#contributing-to-the-fibo-code). My contribution meets all the requirements described there.
- [x] My contribution follows the [principles of best practices for FIBO](https://github.com/edmcouncil/fibo/blob/master/ONTOLOGY_GUIDE.md).
- [x] My changes have been reconciled with latest master and no merge conflicts remain.
- [x] This PR is related to exactly one issue. The issue is referenced by using a GitHub keyword such as "fixes", "closes", or "resolves".
- [x] Hygiene tests have been applied by a PR with "(WIP)" in title.
- [x] The issue has been tested locally using a reasoner (for ontology changes).

### CODE REVIEW

**Review by tahoeblue — APPROVED** [review]

**Review by jfgemski — DISMISSED** [review]

**Review by rivettp — CHANGES_REQUESTED** [review]
The only significant comment for this PR is on the inadequate cardinality e.g. on RateSwap.
Also EquityVolatilitySwap is underspecified.

File: DER/DerivativesContracts/CurrencyContracts.rdf [hunk 1]
```diff
@@ -176,23 +177,31 @@
...
+		<rdfs:subClassOf rdf:resource="&fibo-der-drc-cur;CurrencySwap"/>
 		<rdfs:subClassOf>
 			<owl:Restriction>
 				<owl:onProperty rdf:resource="&fibo-der-drc-swp;hasLeg"/>
-				<owl:someValuesFrom rdf:resource="&fibo-der-drc-cur;CurrencyForward"/>
+				<owl:someValuesFrom rdf:resource="&fibo-der-drc-cur;CurrencySpotContract"/>
 			</owl:Restriction>
 		</rdfs:subClassOf>
+		<rdfs:label xml:lang="en">currency spot forward swap</rdfs:label>
+		<skos:definition xml:lang="en">foreign exchange agreement between two parties involving both an exchange of two currencies on the spot settlement date at a fixed rate that is agreed upon at the inception of the contract covering the exchange; and a reverse exchange of the same two currencies at a later date and at a fixed rate that is agreed upon at the inception of the contract covering the exchange</skos:definition>
```
[the remark this answers is not in the document]
Comment (ElisaKendall) [reply DER/DerivativesContracts/CurrencyContracts.rdf:189]: The dates are associated with the legs of the contract.  The definition of currency spot contract has an execution date.  The forward contract also has settlement terms associated with it that include a settlement date, so we are covered there.

File: DER/DerivativesContracts/Swaps.rdf [hunk 2]
```diff
@@ -239,6 +243,37 @@
 		<skos:definition>floating leg of a swap that depends on some statistical measure of the performance of the underlier</skos:definition>
 	</owl:Class>
 	
+	<owl:Class rdf:about="&fibo-der-drc-swp;RateBasedLeg">
+		<rdfs:subClassOf rdf:resource="&fibo-der-drc-swp;SwapLeg"/>
```
[the remark this answers is not in the document]
Comment (ElisaKendall) [reply DER/DerivativesContracts/Swaps.rdf:247]: SO what are you looking for here?  Do you want the restriction on the leg removed, saying that it is a leg of a swap?  Isn't that part of the definition of "leg"?  The rest of its definition is consistent with the notion of a cashflow, although a leg may or may not take on the structure of a standalone contract.

File: DER/DerivativesContracts/Swaps.rdf [hunk 3]
```diff
@@ -356,7 +391,7 @@
 		</rdfs:subClassOf>
 		<rdfs:label>swap</rdfs:label>
 		<skos:definition>derivative instrument whereby two counterparties agree to exchange periodic streams of cash flows with each other</skos:definition>
-		<fibo-fnd-utl-av:adaptedFrom>ISO 10962:2019 Securities and related financial instruments - Classification of financial instruments (CFI) code</fibo-fnd-utl-av:adaptedFrom>
+		<fibo-fnd-utl-av:adaptedFrom>ISO 10962, Securities and related financial instruments - Classification of financial instruments (CFI) code, Fourth edition, 2019-10</fibo-fnd-utl-av:adaptedFrom>
```
[the remark this answers is not in the document]
Comment (ElisaKendall) [reply DER/DerivativesContracts/Swaps.rdf:394]: done.

File: DER/RateDerivatives/IRSwaps.rdf [hunk 4]
```diff
@@ -356,7 +357,7 @@
 		</rdfs:subClassOf>
 		<rdfs:label xml:lang="en">inflation swap</rdfs:label>
 		<skos:definition xml:lang="en">rate swap in which one party pays an amount calculated using an inflation rate index, and the other party pays an amount calculated using another inflation rate index, or a fixed or floating interest rate</skos:definition>
-		<fibo-fnd-utl-av:adaptedFrom xml:lang="en">ISO 10962, Securities and related financial instruments - Classification of financial instruments (CFI) code, Fourth Edition, 2019.</fibo-fnd-utl-av:adaptedFrom>
+		<fibo-fnd-utl-av:adaptedFrom xml:lang="en">ISO 10962, Securities and related financial instruments - Classification of financial instruments (CFI) code, Fourth Edition, 2019-10.</fibo-fnd-utl-av:adaptedFrom>
```
[the remark this answers is not in the document]
Comment (ElisaKendall) [reply DER/RateDerivatives/IRSwaps.rdf:360]: done

File: DER/RateDerivatives/IRSwaps.rdf [hunk 5]
```diff
@@ -534,6 +535,34 @@
 		<skos:definition>schedule of changes in the notional amount on which interest is paid, comprising the regular sequence of step events</skos:definition>
 	</owl:Class>
 	
+	<owl:Class rdf:about="&fibo-der-rtd-irswp;OvernightIndexSwap">
+		<rdfs:subClassOf rdf:resource="&fibo-der-rtd-irswp;InterestRateSwap"/>
+		<rdfs:subClassOf>
+			<owl:Restriction>
+				<owl:onProperty rdf:resource="&fibo-der-drc-swp;hasLeg"/>
+				<owl:someValuesFrom rdf:resource="&fibo-der-rtd-irswp;OvernightRateIndexLeg"/>
```
[the remark this answers is not in the document]
Comment (ElisaKendall) [reply DER/RateDerivatives/IRSwaps.rdf:543]: No - it's only one leg that is an overnight rate.

File: DER/SecurityBasedDerivatives/EquitySwaps.rdf [hunk 6]
```diff
@@ -145,45 +145,98 @@
...
+		<rdfs:subClassOf rdf:resource="&fibo-der-drc-swp;ReturnSwap"/>
+		<rdfs:subClassOf rdf:resource="&fibo-der-sbd-eqs;EquitySwap"/>
+		<rdfs:subClassOf>
+			<owl:Restriction>
+				<owl:onProperty rdf:resource="&fibo-der-drc-swp;hasReturnLeg"/>
+				<owl:someValuesFrom rdf:resource="&fibo-der-sbd-eqs;EquityReturnLeg"/>
+			</owl:Restriction>
+		</rdfs:subClassOf>
+		<rdfs:label xml:lang="en">equity price return swap</rdfs:label>
+		<skos:definition xml:lang="en">equity swap whose return leg underlier is an equity observable</skos:definition>
```
[the remark this answers is not in the document]
Comment (ElisaKendall) [reply DER/SecurityBasedDerivatives/EquitySwaps.rdf:170]: ignored per your feedback on the DER call

File: DER/SecurityBasedDerivatives/EquitySwaps.rdf [hunk 7]
```diff
@@ -145,45 +145,98 @@
...
-		<rdfs:subClassOf rdf:resource="&fibo-der-drc-swp;TotalReturnSwap"/>
+		<owl:deprecated rdf:datatype="&xsd;boolean">true</owl:deprecated>
+		<owl:equivalentClass rdf:resource="&fibo-der-sbd-eqs;EquityTotalReturnSwap"/>
+	</owl:Class>
+	
+	<owl:Class rdf:about="&fibo-der-sbd-eqs;EquitySwap">
+		<rdfs:subClassOf rdf:resource="&fibo-der-drc-swp;Swap"/>
 		<rdfs:subClassOf rdf:resource="&fibo-der-sbd-sbd;EquityDerivative"/>
+		<rdfs:label xml:lang="en">equity swap</rdfs:label>
+		<skos:definition xml:lang="en">swap whose payments are linked to the change in value of an underlying equity (e.g. shares, basket of equities or index).</skos:definition>
```
[the remark this answers is not in the document]
Comment (ElisaKendall) [reply DER/SecurityBasedDerivatives/EquitySwaps.rdf:196]: added "or its cashflow" to the definition

File: DER/SecurityBasedDerivatives/EquitySwaps.rdf [hunk 8]
```diff
@@ -145,45 +145,98 @@
...
+	</owl:Class>
+	
+	<owl:Class rdf:about="&fibo-der-sbd-eqs;EquityVolatilitySwap">
+		<rdfs:subClassOf rdf:resource="&fibo-der-drc-swp;DispersionSwap"/>
+		<rdfs:subClassOf rdf:resource="&fibo-der-sbd-eqs;EquitySwap"/>
+		<rdfs:label xml:lang="en">equity volatility swap</rdfs:label>
+		<skos:definition xml:lang="en">dispersion swap that is a forward contract on the future realized volatility of its underlying equity observable</skos:definition>
+		<fibo-fnd-utl-av:adaptedFrom xml:lang="en">ISO 10962, Securities and related financial instruments - Classification of financial instruments (CFI) code, Fourth Edition, October 2019</fibo-fnd-utl-av:adaptedFrom>
+		<fibo-fnd-utl-av:explanatoryNote xml:lang="en">An equity volatility swap is a forward swap whose payments depend on the variability of movements in a security or underlying instrument&apos;s price; it is a measure of the amount by which an asset&apos;s price is expected to fluctuate over a given period of time; it is normally measured by the annual standard deviation of daily price changes.</fibo-fnd-utl-av:explanatoryNote>
```
[the remark this answers is not in the document]
Comment (ElisaKendall) [reply DER/SecurityBasedDerivatives/EquitySwaps.rdf:239]: changed per discussion

**Review by jfgemski — APPROVED** [review]

File: DER/DerivativesContracts/Swaps.rdf (hunk 2 above)
[the remark this answers is not in the document]
Comment (rivettp) [reply DER/DerivativesContracts/Swaps.rdf:247]: We should discuss with Jeff independent of this PR

File: DER/DerivativesContracts/Swaps.rdf [hunk 9]
```diff
@@ -239,6 +243,38 @@
...
+		<rdfs:subClassOf>
+			<owl:Restriction>
+				<owl:onProperty rdf:resource="&fibo-der-drc-bsc;hasUnderlier"/>
+				<owl:someValuesFrom rdf:resource="&fibo-der-rtd-rtd;RateBasedObservable"/>
+			</owl:Restriction>
+		</rdfs:subClassOf>
+		<rdfs:subClassOf>
+			<owl:Restriction>
+				<owl:onProperty rdf:resource="&fibo-der-drc-swp;isLegOf"/>
+				<owl:someValuesFrom rdf:resource="&fibo-der-drc-swp;RatesSwap"/>
```
Comment (rivettp) [inline DER/DerivativesContracts/Swaps.rdf:257]: Odd to have the upward restriction, which we don't for other legs. Also the cardinality seems wrong - sure it should be exactly 1 (a given leg cannot be part of many swaps).

File: DER/DerivativesContracts/Swaps.rdf [hunk 10]
```diff
@@ -239,6 +243,38 @@
 		<skos:definition>floating leg of a swap that depends on some statistical measure of the performance of the underlier</skos:definition>
 	</owl:Class>
 	
+	<owl:Class rdf:about="&fibo-der-drc-swp;RateBasedLeg">
+		<rdfs:subClassOf rdf:resource="&fibo-der-drc-swp;SwapLeg"/>
+		<rdfs:subClassOf>
+			<owl:Restriction>
+				<owl:onProperty rdf:resource="&fibo-der-drc-bsc;hasUnderlier"/>
+				<owl:someValuesFrom rdf:resource="&fibo-der-rtd-rtd;RateBasedObservable"/>
```
Comment (rivettp) [inline DER/DerivativesContracts/Swaps.rdf:251]: exactly 1 surely

File: DER/RateDerivatives/IRSwaps.rdf (hunk 5 above)
[the remark this answers is not in the document]
Comment (rivettp) [reply DER/RateDerivatives/IRSwaps.rdf:543]: Ok it's not exactly 2 but it cannot be more than 2.

File: DER/RateDerivatives/IRSwaps.rdf [hunk 11]
```diff
@@ -534,6 +535,34 @@
...
+		<fibo-fnd-utl-av:explanatoryNote xml:lang="en">The interest of the overnight rate portion of the swap is compounded and paid at reset dates. The present value for the leg is determined by either compounding of the overnight rate or by taking the geometric average of the rate over a given period.</fibo-fnd-utl-av:explanatoryNote>
+	</owl:Class>
+	
+	<owl:Class rdf:about="&fibo-der-rtd-irswp;OvernightRateIndexLeg">
+		<rdfs:subClassOf rdf:resource="&fibo-der-rtd-irswp;FloatingInterestRateLeg"/>
+		<rdfs:label>overnight rate index leg</rdfs:label>
+		<owl:equivalentClass>
+			<owl:Restriction>
+				<owl:onProperty rdf:resource="&fibo-fbc-dae-dbt;hasInterestRate"/>
+				<owl:someValuesFrom rdf:resource="&fibo-ind-ir-ir;OvernightRate"/>
```
Comment (rivettp) [inline DER/RateDerivatives/IRSwaps.rdf:559]: exactly 1 for a single leg

File: DER/SecurityBasedDerivatives/EquitySwaps.rdf [hunk 12]
```diff
@@ -145,45 +145,98 @@
...
+		<fibo-fnd-utl-av:adaptedFrom xml:lang="en">ISO 10962, Securities and related financial instruments - Classification of financial instruments (CFI) code, Fifth edition, 2021-06-15</fibo-fnd-utl-av:adaptedFrom>
+		<fibo-fnd-utl-av:explanatoryNote xml:lang="en">A price return equity swap is similar to a total return swap, except that dividends are not passed through to the buyer).</fibo-fnd-utl-av:explanatoryNote>
+	</owl:Class>
+	
+	<owl:Class rdf:about="&fibo-der-sbd-eqs;EquityReturnLeg">
+		<rdfs:subClassOf rdf:resource="&fibo-der-drc-swp;ReturnLeg"/>
+		<rdfs:subClassOf>
+			<owl:Restriction>
+				<owl:onProperty rdf:resource="&fibo-der-drc-bsc;hasUnderlier"/>
+				<owl:someValuesFrom rdf:resource="&fibo-der-sbd-sbd;EquityObservable"/>
```
Comment (rivettp) [inline DER/SecurityBasedDerivatives/EquitySwaps.rdf:180]: exactly 1

File: DER/SecurityBasedDerivatives/EquitySwaps.rdf [hunk 13]
```diff
@@ -145,45 +145,98 @@
...
 		<fibo-fnd-utl-av:explanatoryNote xml:lang="en">The underlier for the leg can be any of (1) dividend stream for a single stock, (2) change in value for a single share, (3) change in value for a basket of shares, (4) change in value for an index, (5) value of a dividend stream for a basket of shares, or (6) comparison of the change in value of a given share or basket or index against something else - for example, a single share against an index, which is the thing you are cross-correlating with the volatility of the share.</fibo-fnd-utl-av:explanatoryNote>
 	</owl:Class>
 	
+	<owl:Class rdf:about="&fibo-der-sbd-eqs;EquityPriceReturnSwap">
+		<rdfs:subClassOf rdf:resource="&fibo-der-drc-swp;ReturnSwap"/>
+		<rdfs:subClassOf rdf:resource="&fibo-der-sbd-eqs;EquitySwap"/>
+		<rdfs:subClassOf>
+			<owl:Restriction>
+				<owl:onProperty rdf:resource="&fibo-der-drc-swp;hasReturnLeg"/>
+				<owl:someValuesFrom rdf:resource="&fibo-der-sbd-eqs;EquityReturnLeg"/>
```
Comment (rivettp) [inline DER/SecurityBasedDerivatives/EquitySwaps.rdf:166]: exactly 1

**Review by rivettp — APPROVED** [review]
I think some cardinality restrictions are not precise enough but OK to defer.

**Pull request merged**